### PR TITLE
[STAGING ONLY]: Azure OpenAI support

### PR DESCRIPTION
**Pull request by trrwilson**
The PR and its topic branch merely track the initial bootstrapping the Azure OpenAI layer; particularly while the component boundaries are still being finalized with codegen client migrations, it's *far* easier to have a one-repo, single-solution inner loop devoid of standard azure-sdk-for-net accoutrements like central package feeds.

It's based off the patterns discussed in the matching PR:
https://github.com/joseharriaga/openai-in-typespec/pull/74

## General approach

- A top-level `AzureOpenAIClient` derives from `OpenAIClient` and serves as the only public entry point into "Azure things" for parity scenarios
- This top-level client has its own pipeline creation and related logic; it plumbs the pipeline, endpoint, and options down into the base library layer via the protected internal constructor
- Each applicable scenario client has an `internal` derived parallel, e.g. `AzureChatClient : ChatClient`
- Individual scenario clients merely add an internal constructor and reimplement protocol methods
- Message creation and other protocol details are handled purely in the current layer; there's no messy inheritance to manage

## Extras

Given the level of protocol method reuse, I wrote a silly-but-helpful `AzureOpenAIPipelineMessageBuilder` that centralizes the repetitive message creation details into a `new ().With().With().Build()` pattern; this is markedly easy to author and keep track of and allows for easy common updates.

### CODE REVIEW

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.Protocol.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+using OpenAI;
+using OpenAI.Audio;
+using System.ClientModel;
+using System.ClientModel.Primitives;
+using System.ComponentModel;
+using System.Text;
+
+namespace Azure.AI.OpenAI.Staging.Audio;
+
+public partial class AzureAudioClient : AudioClient
```
Comment (KrzysztofCwalina) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.Protocol.cs:13]: I thought these would be internal. Only the top level client would be public.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.Protocol.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+{
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public override ClientResult TranscribeAudio(BinaryContent content, string contentType, RequestOptions options = null)
```
Comment (KrzysztofCwalina) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.Protocol.cs:16]: Why do we have to override these methods as opposed to overriding just the CreateTranscriptionRequestMessage method?

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+using System.ClientModel;
+using System.ClientModel.Primitives;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Azure.AI.OpenAI.Staging;
+
+public partial class AzureOpenAIClient : OpenAIClient
+{
+    protected static new internal RequestOptions DefaultRequestOptions => OpenAIClient.DefaultRequestOptions;
+    protected static new internal PipelineMessageClassifier PipelineMessageClassifier200 => OpenAIClient.PipelineMessageClassifier200;
```
Comment (KrzysztofCwalina) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs:29]: why is this needed? Why not use the base property?

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        // To do: proper token conversion from Azure.Core to System.ClientModel
+        throw new NotImplementedException();
+    }
+
+    internal static ApiKeyCredential GetConfiguredKeyCredential()
+    {
+        return new(Environment.GetEnvironmentVariable("AZURE_OPENAI_API_KEY"));
+    }
+    internal static Uri GetConfiguredEndpoint(AzureOpenAIClientOptions options = null)
```
Comment (KrzysztofCwalina) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs:126]: I think we really need to think through how this will integrate with IConfiguration when we add it in the future.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClientOptions.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using OpenAI;
+
+namespace Azure.AI.OpenAI.Staging;
+
+public partial class AzureOpenAIClientOptions : OpenAIClientOptions
+{
+    public string ApiVersion { get; }
```
Comment (KrzysztofCwalina) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClientOptions.cs:10]: This should use the standard Azure SDK way of doing this. See some existing Azure SDK track 2 library, i.e. the ServiceVersion enum.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.Protocol.cs (hunk 1 above)
Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.Protocol.cs:13]: I'll remove these public ones. Instead, I'll update the key-based top-level client constructor to accept a URI:
```
public AzureOpenAIClient(Uri endpoint = null, ApiKeyCredential credential = null, AzureOpenAIClientOptions options = null)
```

The main thing I want to avoid is having the only option for programmatically providing an Azure resource endpoint be using the client options type; that makes sense for the unbranded client, but not for Azure. The above seems to work for that, though, as I can now do this:
```
AzureOpenAIClient topLevelClient = new(new Uri(endpointFromEnvironment), new ApiKeyCredential(keyFromEnvironment));
ChatClient chatClient = topLevelClient.GetChatClient("gpt-35-turbo");
ChatCompletion chatCompletion = chatClient.CompleteChat("hello, world!");
```

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs (hunk 3 above)
Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs:29]: Per Anne's feedback, I've removed the inheritance of these two -- the RequestOptions go away entirely and the classifier becomes per-layer lazy-init property/field pairs.

As the outdated answer, it was just because the `protected internal` access didn't provide visibility to the scenario clients for their protocol method use -- and the main goal was to avoid each and every scenario client having the same thing with initializing these.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs (hunk 4 above)
Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs:126]: I agree. I'll add a similar note to this use of the environment variable as I'd placed on the other in the interim:
```csharp
        // To do: IConfiguration
        configuredEndpoint ??= Environment.GetEnvironmentVariable("AZURE_OPENAI_ENDPOINT");
        if (configuredEndpoint is null)
        {
            throw new InvalidOperationException(
                $"An Azure OpenAI resource endpoint URI must be provided via the client constructor, application "
                + "configuration, or AZURE_OPENAI_ENDPOINT environment variable.");
        }
```

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClientOptions.cs (hunk 5 above)
Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClientOptions.cs:10]: Merged from the existing azure-sdk-for-net library.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace Azure.AI.OpenAI.Staging.Audio;
+
+/// <summary>
+/// The scenario client used for audio operations with the Azure OpenAI service.
+/// </summary>
+/// <remarks>
+/// To retrieve an instance of this type, use the matching method on <see cref="AzureOpenAIClient"/>.
+/// </remarks>
+public partial class AzureAudioClient : AudioClient
```
Comment (annelo-msft) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.cs:15]: Curious: what motivates making these public?

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/Assistants/AzureAssistantClient.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Azure.AI.OpenAI.Staging.FineTuning;
+
+/// <summary>
+/// The scenario client used for assistant operations with the Azure OpenAI service.
+/// </summary>
+/// <remarks>
+/// To retrieve an instance of this type, use the matching method on <see cref="AzureOpenAIClient"/>.
+/// </remarks>
+[Experimental("OPENAI001")]
```
Comment (annelo-msft) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/Assistants/AzureAssistantClient.cs:16]: 🚀

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.cs (hunk 6 above)
Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.cs:15]: Egads, I missed some! They're not intended to be anymore.

It might be nice to not have to create the top-level client, but we agreed we'd consider that later since it's a non-disruptive, additive change.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+{
+    private readonly string _deploymentName;
+    private readonly Uri _endpoint;
+    private readonly string _apiVersion;
+
+    protected internal AzureAudioClient(
```
Comment (annelo-msft) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.cs:21]: Also just curious: what motivates making these protected?  That would enable further client inheritance, but do we have a scenario that would make us want to do that?  I ask just because of the old API design adage -- "you can add it, but you can never take it away again."

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+/// </remarks>
+public partial class AzureOpenAIClient : OpenAIClient
+{
+    private const string azureOpenAIEndpointEnvironmentVariable = "AZURE_OPENAI_ENDPOINT";
+    private const string azureOpenAIApiKeyEnvironmentVariable = "AZURE_OPENAI_API_KEY";
+
+    private readonly AzureOpenAIClientOptions _options;
+
+    internal static PipelineMessageClassifier PipelineMessageClassifier200 => _pipelineMessageClassifier200.Value;
+    internal static Lazy<PipelineMessageClassifier> _pipelineMessageClassifier200
```
Comment (annelo-msft) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs:41]: IIRC using Lazy costs an allocation -- consider making the field private and nullable?  I think there should be examples floating around in generated DPG clients -- lmk if it would be helpful to dig up a link.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+/// For scenario-specific operations, create a corresponding client using the matching method on this type, e.g. <see cref="GetChatClient(string)"/>.
+/// </remarks>
+public partial class AzureOpenAIClient : OpenAIClient
+{
+    private const string azureOpenAIEndpointEnvironmentVariable = "AZURE_OPENAI_ENDPOINT";
+    private const string azureOpenAIApiKeyEnvironmentVariable = "AZURE_OPENAI_API_KEY";
+
+    private readonly AzureOpenAIClientOptions _options;
+
+    internal static PipelineMessageClassifier PipelineMessageClassifier200 => _pipelineMessageClassifier200.Value;
```
Comment (annelo-msft) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs:40]: \<incredibly nit-picky and stylistic\> in our generated code we put these at the bottom of the file, consider following that pattern, but I don't feel strongly about that at all </>

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.cs (hunk 8 above)
Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/Audio/AzureAudioClient.cs]: Very fair point. I don't have any good reason right now; I'll make these plain internal.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs (hunk 9 above)
Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs:41]: Oh, it does? Had no idea. Yes, the pattern was already there for the taking -- I was just being stubborn. I'll replace it in both places.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClientOptions.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+// Licensed under the MIT License.
+
+using OpenAI;
+
+namespace Azure.AI.OpenAI.Staging;
+
+/// <summary>
+/// Defines the scenario-independent, client-level options for the Azure-specific OpenAI client.
+/// </summary>
+public partial class AzureOpenAIClientOptions : OpenAIClientOptions
```
Comment (annelo-msft) [inline .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClientOptions.cs:11]: Nothing to flag here, but as you evolve this forward, if you ever add mutable properties, be sure to use [AssertNotFrozen](https://learn.microsoft.com/en-us/dotnet/api/system.clientmodel.primitives.clientpipelineoptions.assertnotfrozen?view=azure-dotnet-preview#system-clientmodel-primitives-clientpipelineoptions-assertnotfrozen) for them.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs (hunk 10 above)
Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClient.cs:40]: I do want to follow consistent style! Does that apply to all private/internal non-instance fields? I'll move them.

File: .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClientOptions.cs (hunk 11 above)
Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClientOptions.cs:11]: Much appreciated! Is this the right pattern to do it?
```csharp
    internal string Version
    {
        get
        {
            AssertNotFrozen();
            return _version;
        }
    }
    private readonly string _version;
```

Comment (trrwilson) [reply .dotnet/Azure.AI.OpenAI.Staging/Custom/AzureOpenAIClientOptions.cs]: "Mutable" is definitely the key word and I discovered quickly why the assert on a `get` was a bad idea -- but I did leave it in the one place we currently still have a `set` and it seems to work as intended there.
